src/cli/cli_tcp.cpp

```diff
@@ -0,0 +1,520 @@
+    , mEndpointConnected(false)
+    , mSendBusy(false)
+    , mBenchmarkBytesTotal(0)
+    , mBenchmarkLinksLeft(0)
+{
+}
+
+otError TcpExample::ProcessHelp(Arg aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgs);
+
+    for (const Command &command : sCommands)
+    {
+        mInterpreter.OutputLine(command.mName);
+    }
+
+    return OT_ERROR_NONE;
+}
+
+otError TcpExample::ProcessInit(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+    size_t  receiveBufferSize;
+
+    VerifyOrExit(!mInitialized, error = OT_ERROR_ALREADY);
+
+    if (aArgs[0].IsEmpty())
+    {
+        receiveBufferSize = sizeof(mReceiveBuffer);
+    }
+    else
+    {
+        uint32_t windowSize;
+
+        SuccessOrExit(error = aArgs[0].ParseAsUint32(windowSize));
+        VerifyOrExit(aArgs[1].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
+
+        receiveBufferSize = windowSize + ((windowSize + 7) >> 3);
+        VerifyOrExit(receiveBufferSize <= sizeof(mReceiveBuffer) && receiveBufferSize != 0,
+                     error = OT_ERROR_INVALID_ARGS);
+    }
+
+    {
+        otTcpEndpointInitializeArgs endpointArgs;
+
+        memset(&endpointArgs, 0x00, sizeof(endpointArgs));
+        endpointArgs.mEstablishedCallback      = HandleTcpEstablishedCallback;
+        endpointArgs.mSendDoneCallback         = HandleTcpSendDoneCallback;
+        endpointArgs.mReceiveAvailableCallback = HandleTcpReceiveAvailableCallback;
+        endpointArgs.mDisconnectedCallback     = HandleTcpDisconnectedCallback;
+        endpointArgs.mContext                  = this;
+        endpointArgs.mReceiveBuffer            = mReceiveBuffer;
+        endpointArgs.mReceiveBufferSize        = receiveBufferSize;
+
+        SuccessOrExit(error = otTcpEndpointInitialize(mInterpreter.mInstance, &mEndpoint, &endpointArgs));
+    }
+
+    {
+        otTcpListenerInitializeArgs listenerArgs;
+
+        memset(&listenerArgs, 0x00, sizeof(listenerArgs));
+        listenerArgs.mAcceptReadyCallback = HandleTcpAcceptReadyCallback;
+        listenerArgs.mAcceptDoneCallback  = HandleTcpAcceptDoneCallback;
+        listenerArgs.mContext             = this;
+
+        error = otTcpListenerInitialize(mInterpreter.mInstance, &mListener, &listenerArgs);
+        if (error != OT_ERROR_NONE)
+        {
+            IgnoreReturnValue(otTcpEndpointDeinitialize(&mEndpoint));
+            ExitNow();
+        }
+    }
+
+    mInitialized = true;
+
+exit:
+    return error;
+}
+
+otError TcpExample::ProcessDeinit(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+    otError endpointError;
+    otError listenerError;
+
+    VerifyOrExit(aArgs[0].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(mInitialized, error = OT_ERROR_INVALID_STATE);
+
+    endpointError = otTcpEndpointDeinitialize(&mEndpoint);
+    mSendBusy     = false;
+
+    listenerError = otTcpListenerDeinitialize(&mListener);
+    mInitialized  = false;
+
+    SuccessOrExit(error = endpointError);
+    SuccessOrExit(error = listenerError);
```

Comment:
What are the nature of these errors? I'm curious whether we should update `mSendBusy` and/or `mInitialized` when something other than `OT_ERROR_NONE` is returned. Thoughts?

Reply:
In the actual implementation, deinitialization will never fail unless the endpoint/listener wasn't initialized (which would be undefined behavior anyway, since you called a function on uninitialized memory). I just thought it was good practice to check for the error in the CLI tool. Because there isn't a concrete case where this would fail unexpectedly, I wasn't sure how best to handle an error here.

Reply:
Do we envision a use case for the errors? If not, should the API return `void` instead?

Reply:
The reason I left the return value as `otError` was to future-proof the API: perhaps, the implementation will change in some unforeseen way that it can return a meaningful error. Additionally, even in the current implementation, it can return an error in some cases, but only due to programmer error in how the function is used (e.g., if you remove an endpoint that isn't part of the global linked list of endpoints).

I'm happy to change the return value to `void` if you think that that's better.

Reply:
I'm fine with returning error here. Thanks.